# E-commerce Dataset Analysis: Uncovering Data Quality Insights

In [1]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

import warnings
warnings.filterwarnings('ignore')

pd.options.display.max_columns = None
pd.set_option('display.max_rows', 100)
pd.set_option('display.float_format', lambda x: '%.2f' % x)

In [2]:
df = pd.read_csv('Ecommerce_dataset.csv')
df.head()

Order_Date      Time  Aging  Customer_Id  Gender Device_Type  \
0  2018-01-02  10:56:33   8.00        37077  Female         Web   
1  2018-07-24  20:41:37   2.00        59173  Female         Web   
2  2018-11-08  08:38:49   8.00        41066  Female         Web   
3  2018-04-18  19:28:06   7.00        50741  Female         Web   
4  2018-08-13  21:18:39   9.00        53639  Female         Web   

  Customer_Login_type    Product_Category            Product  Sales  Quantity  \
0              Member  Auto & Accessories  Car Media Players 140.00      1.00   
1              Member  Auto & Accessories       Car Speakers 211.00      1.00   
2              Member  Auto & Accessories    Car Body Covers 117.00      5.00   
3              Member  Auto & Accessories    Car & Bike Care 118.00      1.00   
4              Member  Auto & Accessories               Tyre 250.00      1.00   

   Discount  Profit  Shipping_Cost Order_Priority Payment_method  
0      0.30   46.00           4.60         Medium    credit_card  
1      0.30  112.00          11.20         Medium    credit_card  
2      0.10   31.20           3.10       Critical    credit_card  
3      0.30   26.20           2.60           High    credit_card  
4      0.30  160.00          16.00       Critical    credit_card

In [3]:
# Basic dataset information
print("Dataset Shape:", df.shape)
print("\nColumn Names and Data Types:")
print(df.dtypes)
print("\nDataset Info:")
df.info()
print("\nMissing Values Count:")
print(df.isnull().sum())
print("\nMissing Values Percentage:")
print((df.isnull().sum() / len(df) * 100).round(2))

Dataset Shape: (51290, 16)

Column Names and Data Types:
Order_Date              object
Time                    object
Aging                  float64
Customer_Id              int64
Gender                  object
Device_Type             object
Customer_Login_type     object
Product_Category        object
Product                 object
Sales                  float64
Quantity               float64
Discount               float64
Profit                 float64
Shipping_Cost          float64
Order_Priority          object
Payment_method          object
dtype: object

Dataset Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51290 entries, 0 to 51289
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Order_Date           51290 non-null  object 
 1   Time                 51290 non-null  object 
 2   Aging                51289 non-null  float64
 3   Customer_Id          51290 non-null  int64  
 4   Gend

In [4]:
# Summary statistics for numerical columns
print("\nNumerical Column Statistics:")
df.describe()




Numerical Column Statistics:


Aging  Customer_Id    Sales  Quantity  Discount   Profit  \
count 51289.00     51290.00 51289.00  51288.00  51289.00 51290.00   
mean      5.26     58155.76   152.34      2.50      0.30    70.41   
std       2.96     26032.22    66.50      1.51      0.13    48.73   
min       1.00     10000.00    33.00      1.00      0.10     0.50   
25%       3.00     35831.25    85.00      1.00      0.20    24.90   
50%       5.00     61018.00   133.00      2.00      0.30    59.90   
75%       8.00     80736.25   218.00      4.00      0.40   118.40   
max      10.50     99999.00   250.00      5.00      0.50   167.50   

       Shipping_Cost  
count       51289.00  
mean            7.04  
std             4.87  
min             0.10  
25%             2.50  
50%             6.00  
75%            11.80  
max            16.80

In [5]:
# Check unique values for categorical columns
categorical_cols = df.select_dtypes(include=['object']).columns
print("\nUnique Values in Categorical Columns:")
for col in categorical_cols:
    print(f"{col}: {df[col].nunique()} unique values")
    print(f"\nValues: {df[col].unique()[:10]}")  # Show first 10 unique values
    print("-" * 50)


Unique Values in Categorical Columns:
Order_Date: 356 unique values

Values: ['2018-01-02' '2018-07-24' '2018-11-08' '2018-04-18' '2018-08-13'
 '2018-07-09' '2018-05-16' '2018-06-23' '2018-07-29' '2018-07-13']
--------------------------------------------------
Time: 35275 unique values

Values: ['10:56:33' '20:41:37' '08:38:49' '19:28:06' '21:18:39' '21:57:05'
 '13:10:30' '18:29:09' '11:55:02' '19:41:10']
--------------------------------------------------
Gender: 2 unique values

Values: ['Female' 'Male']
--------------------------------------------------
Device_Type: 2 unique values

Values: ['Web' 'Mobile']
--------------------------------------------------
Customer_Login_type: 4 unique values

Values: ['Member' 'Guest' 'New ' 'First SignUp']
--------------------------------------------------
Product_Category: 4 unique values

Values: ['Auto & Accessories' 'Fashion' 'Electronic' 'Home & Furniture']
--------------------------------------------------
Product: 42 unique values

Values:

### Data Overview:
- **Size:** 51,290 transactions with 16 features
- **Data Quality:** Only 6 missing values across all columns (<0.01%)
- **Profit Metrics:** Average profit $70.41 per transaction

### Key Observations:
1. **Product Diversity:** 4 main categories, 42 specific products
2. **Customer Segments:**
    - 2 genders
    - 2 device types (Web, Mobile)
    - 4 login types (Member, Guest, New, First SignUp)
3. **Financial Metrics:**
    - Sales range: $33-$250 per transaction
    - Quantity: 1-5 items per order
    - Discount: 10-50% (avg 30.7%)
    - Profit: $0.50-$167.50 per transaction
4. **Operational Aspects:**
    - 4 orders priorities
    - 5 payment methods
    - Shipping costs: $0.10-$16.80
    - Delivery aging: 1-10.5 days




## Data Cleaning and Transformation: Building a Foundation for Analysis

In any data analysis project, the quality of insights depends entirely on the quality of the underlying data. This section details the systematic approach I took to clean, validate, and transform the e-commerce dataset, ensuring that all subsequent analyses rest on solid ground.


For this e-commerce dataset, I've identified several areas requiring attention:
- Missing values that could skew our analysis
- Data type inconsistencies that might cause calculation errors  
- Formatting issues that could prevent proper grouping
- Business logic validations to ensure data makes sense

In [6]:
# Make a copy of dataframe for cleaning
df_clean = df.copy()

#### Converting Data Types for Analysis

Proper data types are essential for accurate calculations and time-based analysis. Here's standardized data types:

In [7]:
# Create a combined datetime column
df_clean['Order_DateTime'] = pd.to_datetime(df_clean['Order_Date'] + ' ' + df_clean['Time'])

# Convert date to datetime
df_clean['Order_Date'] = pd.to_datetime(df_clean['Order_Date'], format='%Y-%m-%d')

# Extract useful time components for analysis
df_clean['Year'] = df_clean['Order_Date'].dt.year
df_clean['Month'] = df_clean['Order_Date'].dt.month
df_clean['Day'] = df_clean['Order_Date'].dt.day
df_clean['Week'] = df_clean['Order_Date'].dt.isocalendar().week
df_clean['Weekday'] = df_clean['Order_Date'].dt.day_name()
df_clean['Hour'] = df_clean['Order_DateTime'].dt.hour

In [8]:
# Ensure numerical columns are proper data types
numerical_columns = ['Customer_Id', 'Aging', 'Sales', 'Quantity', 'Discount', 'Profit', 'Shipping_Cost']
for column in numerical_columns:
    df_clean[column] = pd.to_numeric(df_clean[column], errors='coerce')

print("Data types after conversion:")
print(df_clean.dtypes)

Data types after conversion:
Order_Date             datetime64[ns]
Time                           object
Aging                         float64
Customer_Id                     int64
Gender                         object
Device_Type                    object
Customer_Login_type            object
Product_Category               object
Product                        object
Sales                         float64
Quantity                      float64
Discount                      float64
Profit                        float64
Shipping_Cost                 float64
Order_Priority                 object
Payment_method                 object
Order_DateTime         datetime64[ns]
Year                            int32
Month                           int32
Day                             int32
Week                           UInt32
Weekday                        object
Hour                            int32
dtype: object


#### Strategy for Missing Values

Given the minimal missing data (less than 0.01% in any column), I've chosen targeted approaches for each column:

**1. Handle missing Discount values**

We will use the discount of a product on a given week of a year.

In [9]:
# Find the product week and year with missing discount
for idx, row in df_clean[df_clean['Discount'].isna()].iterrows():
    missing_discount_product = row['Product']
    missing_discount_week = row['Week']
    missing_discount_year = row['Year']
    
    # Get discount for the same product in the same week
    same_week_discount = df_clean[
        (df_clean['Product'] == missing_discount_product) &
        (df_clean['Week'] == missing_discount_week) &
        (df_clean['Year'] == missing_discount_year) &
        (df_clean['Discount'].notna())
    ]['Discount']
    
    if len(same_week_discount) > 0:
        # Use median of same week discounts
        discount_fill_value = same_week_discount.median()
    else:
        # If no data for same week, then may be there is no discount for the product that week
        discount_fill_value = 0
    
    df_clean.loc[idx, 'Discount'] = discount_fill_value
    print(f"Filled missing discount for {missing_discount_product} (Week: {missing_discount_week}, Year: {missing_discount_year}) with: {discount_fill_value}")

Filled missing discount for Tyre (Week: 31, Year: 2018) with: 0.2


**2. Handle missing Quantity values**

Use most frequent quantity ordered for a product

In [10]:
# Find the products with missing quantities
missing_quantity_products = df_clean[df_clean['Quantity'].isna()]['Product'].unique()
print(f"\nMissing quantity for products: {missing_quantity_products}")

for product in missing_quantity_products:
    # Calculate typical quantity for this product
    frequent_quantity = df_clean[df_clean['Product'] == product]['Quantity'].mode()
    if len(frequent_quantity) > 0:
        fill_quantity = frequent_quantity.iloc[0]
    else:
        # if no mode, use median
        fill_quantity = df_clean[df_clean['Product'] == product]['Quantity'].median()

    df_clean.loc[(df_clean['Quantity'].isna()) & (df_clean['Product'] == product), 'Quantity'] = fill_quantity
    print(f"Filled missing quantity for {product} with: {fill_quantity}")


Missing quantity for products: ['Bike Tyres' 'Car Mat']
Filled missing quantity for Bike Tyres with: 1.0
Filled missing quantity for Car Mat with: 1.0


**3. Handle missing Sales values**

Sales column has unit price of the product. So, find price of product and use it to replace missing value

In [11]:
# Find the product with missing sales
missing_sales_product = df_clean[df_clean['Sales'].isna()]['Product'].iloc[0]
print(f"\nMissing sales for product: {missing_sales_product}")

# Use median sales for this product
sales_fill_value = df_clean[df_clean['Product'] == missing_sales_product]['Sales'].median()
df_clean.loc[df_clean['Sales'].isna(), 'Sales'] = sales_fill_value
print(f"Filled missing sales with: {sales_fill_value}")


Missing sales for product: Car Speakers
Filled missing sales with: 211.0


**4. Handle missing Shipping Cost**
    
Shipping cost generally depends on product (its volume) and distance. 
As there is no information related to customer location in the dataset, we will use median shipping cost for a product.

In [12]:
missing_shipping_product = df_clean[df_clean['Shipping_Cost'].isna()]['Product'].iloc[0]
print(f"\nMissing shipping cost for product: {missing_shipping_product}")

# Use median shipping cost for this product
shipping_fill_value = df_clean[df_clean['Product'] == missing_shipping_product]['Shipping_Cost'].median()
df_clean.loc[df_clean['Shipping_Cost'].isna(), 'Shipping_Cost'] = shipping_fill_value
print(f"Filled missing shipping cost with: {shipping_fill_value}")


Missing shipping cost for product: Tyre
Filled missing shipping cost with: 15.25


**5. Handle missing Aging**
    
Use median aging value based on Order_Priority

In [13]:
missing_aging_Order_Priority = df_clean[df_clean['Aging'].isna()]['Order_Priority'].iloc[0]
print(f"\nMissing Aging for Order Priority: {missing_aging_Order_Priority}")

# Use median aging value based on Order_Priority
aging_fill_value = df_clean[df_clean['Order_Priority'] == missing_aging_Order_Priority]['Aging'].median()
df_clean.loc[df_clean['Aging'].isna(), 'Aging'] = aging_fill_value
print(f"Filled missing Aging with: {aging_fill_value}")


Missing Aging for Order Priority: High
Filled missing Aging with: 5.0


**6. Handle missing Order_Priority**

Replace with 'Not Specified'

In [14]:
df_clean['Order_Priority'].fillna('Not Specified', inplace=True)

### Cleaning Text Fields

Text inconsistencies can cause grouping errors and misclassification. I'm standardizing all text fields to ensure consistency:

In [15]:
# Standardize all text columns
text_columns = ['Gender', 'Device_Type', 'Customer_Login_type', 'Product_Category', 
                'Product', 'Order_Priority', 'Payment_method']

for column in text_columns:
    df_clean[column] = df_clean[column].str.strip()
    df_clean[column] = df_clean[column].str.title() # Standardize capitalization

print("\nStandardized text values:")
for column in text_columns:
    print(f"\n{column} unique values: {df_clean[column].nunique()}")
    print(df_clean[column].value_counts().head())


Standardized text values:

Gender unique values: 2
Gender
Male      28138
Female    23152
Name: count, dtype: int64

Device_Type unique values: 2
Device_Type
Web       47632
Mobile     3658
Name: count, dtype: int64

Customer_Login_type unique values: 4
Customer_Login_type
Member          49097
Guest            1993
First Signup      173
New                27
Name: count, dtype: int64

Product_Category unique values: 4
Product_Category
Fashion               25646
Home & Furniture      15438
Auto & Accessories     7505
Electronic             2701
Name: count, dtype: int64

Product unique values: 42
Product
Suits           2332
T - Shirts      2332
Fossil Watch    2332
Shirts          2332
Jeans           2332
Name: count, dtype: int64

Order_Priority unique values: 5
Order_Priority
Medium           29433
High             15499
Critical          3932
Low               2424
Not Specified        2
Name: count, dtype: int64

Payment_method unique values: 5
Payment_method
Credit_Card    381

### Creating Enhanced Business Metrics

**Building Accurate Financial Metrics:** Based on data investigation, I'm creating corrected financial columns that reflect true business performance:

In [16]:
# Renaming for clarity
df_clean.rename(columns={'Sales': 'Unit_Price', 'Profit': 'Unit_Profit'}, inplace=True)

# Create corrected financial columns
df_clean['Total_Sales'] = df_clean['Unit_Price'] * df_clean['Quantity']
df_clean['Total_Profit'] = df_clean['Unit_Profit'] * df_clean['Quantity']

In [17]:
# Create unique order identifiers
df_clean['Order_Id'] = (
    df_clean['Customer_Id'].astype(str) + '_' +
    df_clean['Order_Date'].dt.strftime('%Y-%m-%d') + '_' +
    df_clean['Time']
)

df_clean['Order_Id'].head()

0    37077_2018-01-02_10:56:33
1    59173_2018-07-24_20:41:37
2    41066_2018-11-08_08:38:49
3    50741_2018-04-18_19:28:06
4    53639_2018-08-13_21:18:39
Name: Order_Id, dtype: object

In [18]:
# Calculate additional business metrics
df_clean['Profit_Margin'] = (df_clean['Total_Profit'] / df_clean['Total_Sales'] * 100).round(2)
df_clean['Discount_Amount'] = df_clean['Total_Sales'] * df_clean['Discount']
df_clean['Net_Sales'] = df_clean['Total_Sales'] - df_clean['Discount_Amount']
df_clean['Average_Item_Value'] = df_clean['Total_Sales'] / df_clean['Quantity']

# Create customer value metrics
df_clean['Revenue_After_Shipping'] = df_clean['Net_Sales'] - df_clean['Shipping_Cost']
df_clean['Net_Profit_Margin'] = (df_clean['Total_Profit'] / df_clean['Net_Sales'] * 100).round(2)

#### Validating Transformations

Quality assurance is crucial. Let's verify that transformations maintain data integrity and make business sense:

In [19]:
# Validation checks
validation_results = []

# Check 1: No negative Quantities
neg_quantity = df_clean[df_clean['Quantity'] < 0]
validation_results.append(("Negative Quantities", len(neg_quantity)))

# Check 2: No negative Prices
neg_price = df_clean[df_clean['Unit_Price'] < 0]
validation_results.append(("Negative Prices", len(neg_price)))

# Check 3: Discount range validation (0-100%)
invalid_discount = df_clean[(df_clean['Discount'] < 0) | (df_clean['Discount'] > 1)]
validation_results.append(("Invalid Discounts", len(invalid_discount)))

# Check 4: Order_ID uniqueness
duplicate_orders = df_clean['Order_Id'].duplicated().sum()
validation_results.append(("Duplicate Order IDs", duplicate_orders))

# Check 5: Date range consistency
date_range = f"{df_clean['Order_Date'].min().date()} to {df_clean['Order_Date'].max().date()}"
validation_results.append(("Date Range", date_range))

# Display validation results
validation_df = pd.DataFrame(validation_results, columns=['Check', 'Result'])
print("\nData Validation Results:")
print(validation_df)


Data Validation Results:
                 Check                    Result
0  Negative Quantities                         0
1      Negative Prices                         0
2    Invalid Discounts                         0
3  Duplicate Order IDs                         0
4           Date Range  2018-01-01 to 2018-12-30


In [20]:
# Verify profit scaling with example
example_validation = df_clean[df_clean['Product'] == 'Tyre'].groupby('Quantity').agg({
    'Unit_Profit': 'mean',
    'Total_Profit': 'mean',
    'Order_Id': 'count'
}).round(2)

print("\nProfit Scaling Validation (Tyre Product):")
print(example_validation)


Profit Scaling Validation (Tyre Product):
          Unit_Profit  Total_Profit  Order_Id
Quantity                                     
1.00           153.51        153.51       484
2.00           154.87        309.74        77
3.00           150.17        450.51        59
4.00           138.54        554.14       157
5.00           133.36        666.81       116


#### Summary of Data Cleaning Process

The systematic data cleaning has transformed the raw dataset into a reliable foundation for business analysis. Here's what is accomplished:

In [21]:
# Create a comprehensive summary
summary_stats = {
    'Total Records': len(df_clean),
    'Unique Customers': df_clean['Customer_Id'].nunique(),
    'Unique Products': df_clean['Product'].nunique(),
    'Product Categories': df_clean['Product_Category'].nunique(),
    'Date Range': f"{df_clean['Order_Date'].min().date()} to {df_clean['Order_Date'].max().date()}",
    'Total Revenue': f"${df_clean['Total_Sales'].sum():,.2f}",
    'Total Profit': f"${df_clean['Total_Profit'].sum():,.2f}",
    'Average Order Value': f"${df_clean['Total_Sales'].mean():,.2f}",
    'Average Profit Margin': f"{df_clean['Profit_Margin'].mean():.1f}%"
}

summary_df = pd.DataFrame(list(summary_stats.items()), columns=['Metric', 'Value'])
print("\nCleaned Dataset Summary:")
print(summary_df)


Cleaned Dataset Summary:
                  Metric                     Value
0          Total Records                     51290
1       Unique Customers                     38997
2        Unique Products                        42
3     Product Categories                         4
4             Date Range  2018-01-01 to 2018-12-30
5          Total Revenue            $19,636,297.00
6           Total Profit             $8,577,505.60
7    Average Order Value                   $382.85
8  Average Profit Margin                     42.5%


In [23]:
# Save the cleaned dataset
#df_clean.to_csv('ecommerce_cleaned.csv', index=False)
df_clean.to_parquet('ecommerce_cleaned.parquet', index=False)
print("\nCleaned dataset saved as 'ecommerce_cleaned.parquet'")
print(f"Final dataset shape: {df_clean.shape}")


Cleaned dataset saved as 'ecommerce_cleaned.parquet'
Final dataset shape: (51290, 32)
